cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/operationrunner/InMemorySourceControlOperationRunner.java

```diff
@@ -207,15 +268,46 @@ private PushAppResponse writeAppDetailAndPush(RepositoryManager repositoryManage
       );
     }
 
-    LOG.debug("Wrote application configs for {} in file {}", appToPush.getName(), appRelativePath);
+    LOG.debug("Wrote application configs for {} in file {}", appToPush.getName(),
+        appRelativePath);
+  }
+
+  /**
+   * Atomic operation of writing commit and push, return push responses.
+   *
+   * @param <T>  a type implementing {@link Supplier} representing a file to be pushed (with path)
+   * @param repositoryManager {@link RepositoryManager} to conduct git operations
+   * @param commitMeta {@link CommitMeta} from user input
+   * @param commitFiles an {@link Iterator} of objects of the generic type T.
+   * @param consumer a {@link BiConsumer} that accepts an object of the generic type T and a fileHash
+   *
+   * @throws NoChangesToPushException if there's no effective change between applications in namespace
+   *     and git repository
+   * @throws SourceControlException for failures while writing config file or doing git
+   *     operations
+   */
+  //TODO: CDAP-20371, Add retry logic here in case the head at remote moved while we are doing push
+  private <T extends Supplier<Path>> void commitAndPush(
```

Comment:
the previous comment was about changing the RepositoryManager.commitAndPush() method to take the Iterator and BiConsumer. That way this method doesn't need to create the extra collections like the current pathMapping Map.

Reply:
The `RepositoryManager.commitAndPush` need to construct a set of paths from the iterator anyway, as the set is necessary to efficiently filter the files in the TreeWalk. Also, makes sense for the `RepositoryManager.commitAndPush` to return a Map of filepaths to hashes (as it's one operation, that adds multiple files to a commit and pushes) instead of taking a consumer to pass the results one by one. Therefore, updating this method did not improve readability much.

However, the suggested change made perfect sense for the `commitAndPush` method in the runner, as it simplified the implementation of this method.

Reply:
The original comment was mostly about avoiding all of these extra collections, as they add time, memory, and complexity. I don't think it's actually good that commitAndPush returns a Map, as it almost always forces the caller to eventually transform it into some other type of collection.

If we want it to return information about what it committed instead of passing in a consumer, a better approach would be for it to return a `Collection<O>` and to take a `BiFunction<T, String, O>`, where `O` would be a `PushAppResponse` in this class. I think the current implementation of commitAndPush would make that difficult, but see the other comment I made about getting the file hash right when adding the file instead of walking through the tree after committing.